spatialprofilingtoolbox/apiserver/app/validation.py

```diff
@@ -75,6 +74,16 @@ async def valid_channel_list_positives(positive_marker: ChannelList) -> list[str
 async def valid_channel_list_negatives(negative_marker: ChannelList) -> list[str]:
     return valid_channel_list(negative_marker)
 
+
+squidpy_feature_classnames = ('neighorhood enrichment', 'co-occurrence', 'ripley')
```

Comment:
Are you only allowing the user to retrieve one of the three Squidpy features at a time? The three of them share the initialized neighborhood calculation that would have to be repeated if it isn't saved, and I don't see a reason why a user would prefer just one instead of all three of them, since they share the same input parameters. (Although it could be beneficial to return them sequentially as they're finished calculating.)

Reply:
Hm OK, I didn't think about how that overhead might be a dominant expenditure of computation time.
This will make the squidpy metrics break the mold a little, providing multiple features and computing multiple features on every query.

Reply:
There are numerous places where one occurrence of query handling is associated with one feature specification, so these will have to be rewritten. For exapmle `PendingProvider.get_metrics`, which was supposed to be the generalized implementation.

Reply:
I discussed this a little in the other threads:
https://github.com/nadeemlab/SPT/pull/183#discussion_r1287726451 (*... supposed to exactly tell what feature values were intended to be computed ...*)
https://github.com/nadeemlab/SPT/pull/183#discussion_r1279753401 (*...I don't think this counts as a feature at all.*)

I have been sticking to the original planned goals of #120, including:

*The resulting features should get loaded into the* `feature_specifier/feature_specification/quantitative_feature_value` *system.*

The non-scalar return values you are referring to are not yet summarized, so they are not first-class features according to my "rigid" definition, which requires the simple/univariate aspect.

But that is not all. My definition also stipulates that the derivation method should be carefully described, and we still lack this as well.

My strategy for achieving explainable outputs is (1) identify representative summary features for each unstructured bundle, (2) research the implementation to understand what the summary feature of the unstructured bundle means.

This is what I plan to do after refactoring the squidpy-related modules in order to be feature-oriented.

Reply:
This turned out to be pretty easy for the neighborhood enrichment test.
Here is an example of the return value:
```json
 {
    "zscore": [
        [0.5406386960248688, -0.19020676295202088, 0.4386328402677137],
        [-0.3693187058572753, 0.3503620265496884, -0.4386328402677137],
        [-1.0974813402382708, -0.992579671981139, null]
    ],
    "count": [
        [104, 142, 3],
        [141, 204, 3],
        [1, 2, 0]
    ]
}
```
I checked the source code, and here is my draft description:

```txt
For two given cell phenotypes (first and second specifier), the
z-score of the number of occurrences of neighbor relations
between two cells of the given phenotypes, the z-score computed
in reference to a bootstrapped empirical distribution constructed
using 1000 random permutations of the cell set with labels fixed.
```

The "non-scalar" return value in this case is mainly non-scalar because we have not restricted our attention to the 2 phenotypes which are the main arguments. (The matrix of z-scores is a z-score for each pair of phenotypes). For our purposes I think we should convert this z-score back to a probability value (and change "z score" => "p value" in the description), for a more direct interpretation of the result: (an esimate of) the fraction of the number of times that the random bootstrapping value was higher than the observed count.

In addition to z-scores, "counts" are returned. This is essentially the "test statistic" value corresponding to the p-value. I don't think it is essential to record this and report it in the response. If it turns out to be more important than I thought, we can always make a new feature for it.

Reply:
Note that to get the actual fraction (as opposed to "an estimate of"), we would need to slightly change the squidpy implementation. Specifically this line:

https://github.com/scverse/squidpy/blob/8cd23551e5e14a95ae769d426c33813e5d889179/src/squidpy/gr/_nhood.py#L188C5-L188C62

```py
zscore = (count - perms.mean(axis=0)) / perms.std(axis=0)
```

It would need to be something like
```py
fraction = sum([1 for _count in perms if _count > count]) / len(perms)
```

Reply:
Are you suggesting we fork and maintain a separate branch of squidpy for SPT?

Reply:
The estimate is good enough for our purposes, but when we are ready we could make a PR and request a feature that makes this information available optionally.

Reply:
I made a first pass at the same task for the co-occurrence scores. Even after studying the squidpy implementation, I don't have perfect confidence that I have understood the outputs correctly. I tested it with varying parameters (mainly `interval`), and I have a reasonable theory about the output format. I think it is a 3-axis array, indexable respectively by: cluster, cluster, and distance interval. If they are ordered as expected, I think I am extracting the correct value.
As implemented now, this metric depends on 3 specifiers: a phenotype, a second phenotype, and a radius. I had to add a radius parameter throughout, since previously we assumed that only a phenotype list would be required to specify one of these features.
At the UI level we will likely use a reasonable default for this radius value, but I think it should still be configurable by the (power) user.